### PR TITLE
Refocussed the API overview section.

### PR DESCRIPTION
**Pull request by jeromekelleher**
I've mostly rewritten the "Overview of the API" section and recast it into a "Implementation Overview". I think it's useful to give a bird's eye view of what's going on in the implementations without getting too much into the gorey details.

I've probably got some stuff wrong here in terms of the simulator implementations, so please feel free to correct!

Also, I'm not sure if "proof-of-concept" is how we want to refer to the implementations?

### CODE REVIEW

File: forwards_paper.tex [hunk 1]
```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
...
 
-These tools are implemented through the \msprime{} Python API,
-which provides a powerful platform for working with tree topology and mutation data.
-The new portions of \msprime{} that we cover here
-are dedicated to tree sequence input and output using simple tables of data,
-so we refer to this as the `Tables API'.
-The four key tables: nodes, edges, sites and mutations, are described above.
+The facilities for working with succinct tree sequences are implemented as part
+of the \msprime\ Python API, which provides a powerful platform for processing
+tree topology and mutation data. The new portions of \msprime{} that we discuss
```
Comment (ashander) [inline forwards_paper.tex:910]: when and where do you use `\msprime\` vs `\msprime{}`? need to catch up on .tex tricks

File: forwards_paper.tex [hunk 2]
```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
...
+For small scale data and debugging purposes, a simple text based format is also supported.
+
+The \msprime\ Python Tables API provides a general purpose toolkit for importing
+and processing succinct tree sequences. Interoperation with Python simulators
+is then straightforward. For the proof-of-concept implementations discussed
+here, we took two different approaches to doing this. In the first method,
+based on \simupop{}, the simulation engine invokes callbacks at critical points
+of the simulation, and we infer nodes and edges from the information that is
+provided. Rows are appended to the tables one-by-one, and the tables are
+periodically sorted and simplified to control memory usage. The second method we developed uses
```
Comment (ashander) [inline forwards_paper.tex:941]: Nice description

File: forwards_paper.tex [hunk 3]
```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
...
+\texttt{pybind11} (\url{https://github.com/pybind/pybind11/}) to inferface
+with the \fwdpp{} \cpp{} API \citep{fwdpp}. No modifications were
+required to the \fwdpp{} code base; rather, we simply need to bookkeep parent/offspring labels,
+and perform simple processing of the recombination breakpoints from each mating
+event to generate node and edge data. This information is then periodically copied
+to the \msprime\ Tables API, where it is sorted and simplified.
```
Comment (ashander) [inline forwards_paper.tex:947]: Some of it isn't even copied, I believe, due to zero-copy access mentioned above. @molpopgen can confirm

File: forwards_paper.tex [hunk 4]
```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
...
+
+While these methods work well, and indeed result in significant performance
+gains over non-ancestry based simulation, they are far from optimal. Because
+forwards time simulation is highly CPU intensive, most simulation engines
+are written in C or C++ out of necessity. Therefore, the simplest and most
+efficient approach would be to call the \msprime\ C library directly as part of
+the core simulation logic, in the manner outlined in Algorithm W. If designed
+in this way, the simulation engine would not need to keep copies of the node
+and edge information and would not need to infer edges;
```
Comment (ashander) [inline forwards_paper.tex:956]: "infer edges" isn't true in the c++ implementation so is a bit confusing to state as a general issue. Should remove as simuPOP details will be in appendix if present at all.

File: forwards_paper.tex [hunk 5]
```diff
@@ -903,59 +903,80 @@ \subsubsection*{Estimates of run-time complexity}
...
-Our implementation also uses \texttt{pybind11}
-to create a Python package extending \texttt{fwdpy11}
-(a Python package based on \fwdpp{}; \url{https://github.com/molpopgen/fwdpy11}),
-% (A neat aside, that we may wish to skip here, is that it may be possible to use the tables
-% API from a command-line C/\cpp{} application via an embedded Python interpreter.)
+For small scale data and debugging purposes, a simple text based format is also supported.
+
+The \msprime\ Python Tables API provides a general purpose toolkit for importing
+and processing succinct tree sequences. Interoperation with Python simulators
+is then straightforward. For the proof-of-concept implementations discussed
```
Comment (ashander) [inline forwards_paper.tex:936]: why not just this "the implementations described"?

**Review by ashander — APPROVED** [review]
Looks good.  A few suggestions

**Comment by ashander** [comment]
Unless @petrelharp objects I will merge this in while doing the simupop-related edits

File: forwards_paper.tex (hunk 3 above)
Comment (jeromekelleher) [reply forwards_paper.tex:947]: It's still copied into msprime's memory --- the zero-copy Kevin's referring to is the memory view of C++ structs available in numpy.

File: forwards_paper.tex (hunk 5 above)
Comment (jeromekelleher) [reply forwards_paper.tex:936]: Yes, good idea. We can revisit this.

File: forwards_paper.tex (hunk 1 above)
Comment (jeromekelleher) [reply forwards_paper.tex:910]: It doesn't matter --- I just use the trailing backslash out of habit.

File: forwards_paper.tex (hunk 3 above)
Comment (ashander) [reply forwards_paper.tex:947]: OK gotcha. A python user can in principle use a view of msprime data without a copy; such user can do the same with code from a forward simulator that uses the NumPy API. But getting data from fwd simulator into msprime will always involve a copy.

Comment (jeromekelleher) [reply forwards_paper.tex:947]: Yes, exactly. Ideally, we'd like to close this gap between the forward simulator and the tables API.

File: forwards_paper.tex (hunk 4 above)
Comment (molpopgen) [reply forwards_paper.tex:956]: I'd have to reread pybind11 doc, but my understanding is that I am not using the numpy api, actually.  I'm using the buffer protocol and registering the stored struct as numpy dtypes.  The conversion to a numpy record arrays is zero copy due to buffer protocol.

Pybind11's entire numpy support works without ever including numpy headers nor linking to it.  Black magic, for sure.

**Pull request merged**